registry/handlers/helpers.go

```diff
@@ -64,3 +74,35 @@ func copyFullPayload(ctx context.Context, responseWriter http.ResponseWriter, r
 
 	return nil
 }
+
+// checkForClientDisconnection is a generic function which checks if a HTTP request for a given client has been closed
+// and if it has returns a typed client disconnection event
+func checkForClientDisconnection(w http.ResponseWriter, r *http.Request) error {
+	var body = r.Body
+	clientClosed := r.Context().Done()
+	bodyLen, err := body.Read([]byte{})
+	if clientClosed != nil && (err != nil || (r.ContentLength > 0 && int64(bodyLen) < r.ContentLength)) {
+		select {
+		case <-clientClosed:
+			w.WriteHeader(499)
+			return ErrorClientDisconnected{}
+		default:
+		}
+	}
+	return nil
+}
+
+// handleDisconnectionEvent is a utility abstraction for checking client disconnection events and ensures the correct
+// 499 error is surfaced to the client. As this logic is used in multiple places this prevents the handling logic
+// from being duplicated
+func handleDisconnectionEvent(ctx *Context, w http.ResponseWriter, r *http.Request) ([]error, bool) {
+	handled := false
+	disconnected := checkForClientDisconnection(w, r)
```

Comment:
Couldn't we just place the body of `checkForClientDisconnected` here? Is it separated into its own function for the sake of testing? I ask because it seems weird to do a type assertion at line 102 just for sake of separating the `if clientClosed ...` logic into a separate function.

Reply:
I think at one point this was separated because of where I had placed the errors too low inside the storage driver. As I hoisted things around during the code review I don't think its required any more. I'll merge for now and maybe we can clean this up another time as I know the code in its current form works to avoid another review cycle